File: src/main/java/leetcode/add_two_numbers/AddTwoNumber.java
package leetcode.add_two_numbers;

import java.util.LinkedList;
import java.util.List;

public class AddTwoNumber {

    public static void main(String[] args) {
        System.out.println(addTwoNumber(new LinkedList<>(List.of(2, 4, 3)), new LinkedList<>(List.of(5, 6, 4))));
        System.out.println(addTwoNumber(new LinkedList<>(List.of(5, 6, 4)), new LinkedList<>(List.of(2, 4, 3, 3))));
    }

    public static LinkedList<Integer> addTwoNumber(LinkedList<Integer> first, LinkedList<Integer> second) {
        LinkedList<Integer> result = new LinkedList<>();
        /*instead of finding the biggest one we can use iterator*/
        var small = first.size() > second.size() ? second : first;
        var big = first.size() > second.size() ? first : second;

        int carry = 0;
        for (int i = 0; i < big.size(); i++) {
            int value = big.get(i) + (i >= small.size() ? 0 : small.get(i)) + carry;
            carry = value / 10;
            result.add(value % 10);
        }
        if (carry > 0)
            result.add(carry);
        return result;
    }

}

File: src/main/java/leetcode/add_two_numbers/AddTwoNumber.kt
package leetcode.add_two_numbers

import java.util.*

fun main() {
    println(addTwoNumber(LinkedList(listOf(2, 4, 3)), LinkedList(listOf(5, 6, 4))))
    println(addTwoNumber(LinkedList(listOf(5, 6, 4)), LinkedList(listOf(2, 4, 3, 3))))
}

fun addTwoNumber(first: LinkedList<Int>, second: LinkedList<Int>): LinkedList<Int>? {
    val result = LinkedList<Int>()
    /*instead of finding the biggest one we can use iterator*/
    val small = if (first.size > second.size) second else first
    val big = if (first.size > second.size) first else second
    var carry = 0
    for (i in big.indices) {
        val value = big[i] + (if (i >= small.size) 0 else small[i]) + carry
        carry = value / 10
        result.add(value % 10)
    }
    if (carry > 0) result.add(carry)
    return result
}